### PR TITLE
runfix: Harmonize gracePeriod computation for new and renewed devices

### PR DESCRIPTION
**Pull request by atomrc**
## Description

Currently both the initial enrollment and renewals should follow the same timer logic. 
A enroll/renew modal should be shown depending on some grace period depending on the device state and the grace period set on team settings (see https://wearezeta.atlassian.net/wiki/spaces/ENGINEERIN/pages/961478669/Use+case%3A+renew+end-to-end+identity+certificate). 

This PR refactors how this timer is computed and use the same timer for initial enrollment and renewal. 

It also fixes the computation of the softLock state of the app (now the enrollment class is responsible for computing the status of the device)

### Creating first certificate

https://github.com/wireapp/wire-webapp/assets/1090716/bd9a2a41-3c18-402d-b770-3e4035dee47f

### Renewing certificate (not snoozable)

https://github.com/wireapp/wire-webapp/assets/1090716/befda273-2c2c-49d2-b080-890dad0e41c5

### Renewing certificate (snoozable)

https://github.com/wireapp/wire-webapp/assets/1090716/0dd9a2b2-2d2d-4181-b9da-9fc237517495

## Checklist

- [x] PR has been self reviewed by the author;
- [x] Hard-to-understand areas of the code have been commented;
- [x] If it is a core feature, unit tests have been added;

### CODE REVIEW

File: src/script/E2EIdentity/E2EIdentityEnrollment.ts [hunk 1]
```diff
@@ -117,107 +112,69 @@ export class E2EIHandler extends TypedEventEmitter<Events> {
...
     await this.coreE2EIService.initialize(discoveryUrl);
 
-    this.emit('initialized', {enrollmentConfig: this.config});
+    if (await this.coreE2EIService.isEnrollmentInProgress()) {
+      // If we have an enrollment in progress, we can just finish it (meaning we are coming back from an idp redirect)
+      await this.enroll();
+    } else if (await isFreshMLSSelfClient()) {
+      // When the user logs in to a new device in an environment that has e2ei enabled, they should be forced to enroll
+      await this.startEnrollment(ModalType.ENROLL, false);
+    }
```
Comment (atomrc) [inline src/script/E2EIdentity/E2EIdentityEnrollment.ts]: This is a new behavior, when we initialize and we either have:
- a new device
- an enrollment in progress 

we automatically kick off the enrollement process

File: src/script/E2EIdentity/E2EIdentityEnrollment.ts [hunk 2]
```diff
@@ -117,107 +112,69 @@ export class E2EIHandler extends TypedEventEmitter<Events> {
...
-      // If the client already has a certificate, we don't need to start the enrollment
-      return;
-    }
-    return this.startEnrollment(ModalType.ENROLL);
-  }
+  /**
+   * Will initiate the timer that will regularly prompt the user to enroll (or to renew the certificate if it is about to expire)
+   * @returns the delay under which the next enrollment/renewal modal will be prompted
+   */
+  public async startTimers() {
```
Comment (atomrc) [inline src/script/E2EIdentity/E2EIdentityEnrollment.ts:143]: This method is the main change here. 

Those timers are handling every single future enrollment triggers (other than the initial enrollment for new devices)

File: src/script/E2EIdentity/E2EIdentityEnrollment.ts [hunk 3]
```diff
@@ -117,107 +112,69 @@ export class E2EIHandler extends TypedEventEmitter<Events> {
...
+    // We store the first time the user was prompted with the enrollment modal
+    const e2eActivatedAt = SnoozableTimerStore.get.e2eiActivatedAt() || Date.now();
+    SnoozableTimerStore.store.e2eiActivatedAt(e2eActivatedAt);
 
-  public async attemptRenewal(): Promise<void> {
+    const timerKey = 'enrollmentTimer';
     const identity = await getActiveWireIdentity();
+    const {firingDate, isSnoozable} = getEnrollmentTimer(identity, e2eActivatedAt, this.config.gracePeriodInMs);
```
Comment (atomrc) [inline src/script/E2EIdentity/E2EIdentityEnrollment.ts:150]: This `snoozable` is the single source of truth for `softLock`. We do not try to compute it anywhere else

File: src/script/E2EIdentity/SnoozableTimer/SnoozableTimer.ts
Comment (atomrc) [inline src/script/E2EIdentity/SnoozableTimer/SnoozableTimer.ts:1]: This snoozable timer is now gone in favor of a simple `TaskScheduler` used in the enrollment class

File: src/script/main/app.ts [hunk 4]
```diff
@@ -398,14 +398,6 @@ export class App {
       const e2eiHandler = await configureE2EI(this.logger, teamFeatures);
       configureDownloadPath(teamFeatures);
 
-      if (e2eiHandler) {
-        /* We first try to do the initial enrollment (if the user has not yet enrolled)
-         * We need to enroll before anything else (in particular joining MLS conversations)
-         * Until the user is enrolled, we need to pause loading the app
-         */
-        await e2eiHandler.attemptEnrollment();
-      }
```
Comment (atomrc) [inline src/script/main/app.ts:407]: This is handled by the `instantiate` method now

File: src/script/main/app.ts [hunk 5]
```diff
@@ -537,16 +529,9 @@ export class App {
...
       this.logger.info(`App loaded in ${Date.now() - startTime}ms`);
 
-      if (e2eiHandler) {
-        // At the end of the process (once conversations are loaded and joined), we can check if we need to renew the user's certificate
-        try {
-          await e2eiHandler.attemptRenewal();
-        } catch (error) {
-          this.logger.error('Failed to renew user certificate: ', error);
-        }
-      }
```
Comment (atomrc) [inline src/script/main/app.ts:549]: Timers are now responsible for triggering a renewal if needs be

File: src/script/E2EIdentity/E2EIdentityEnrollment.ts [hunk 6]
```diff
@@ -17,60 +17,48 @@
...
 };
 
-const historyTimeMS = 28 * TimeInMillis.DAY; //HT
 export class E2EIHandler extends TypedEventEmitter<Events> {
   private logger = getLogger('E2EIHandler');
   private static instance: E2EIHandler | null = null;
   private readonly core = container.resolve(Core);
   private readonly userState = container.resolve(UserState);
-  private config?: EnrollmentConfig;
+  #config?: EnrollmentConfig;
```
Comment (przemvs) [inline src/script/E2EIdentity/E2EIdentityEnrollment.ts:62]: Wow! Starting using new thins? :)

File: src/script/E2EIdentity/E2EIdentityEnrollment.ts [hunk 7]
```diff
@@ -117,107 +112,68 @@ export class E2EIHandler extends TypedEventEmitter<Events> {
...
-    const deviceIdentityStatus = identity.status;
-    switch (deviceIdentityStatus) {
-      case MLSStatuses.REVOKED:
-        return false;
-      default:
-        return true;
-    }
+    const fireIn = firingDate - Date.now();
+    this.logger.info(`Enrollment timer started {snoozable: ${isSnoozable}, fireAt: ${new Date(firingDate)}}`);
+    return fireIn;
```
Comment (przemvs) [inline src/script/E2EIdentity/E2EIdentityEnrollment.ts]: Can't by like that? :) You're not using fireIn in another places ;)

```suggestion
    this.logger.info(`Enrollment timer started {snoozable: ${isSnoozable}, fireAt: ${new Date(firingDate)}}`);
    return firingDate - Date.now();
```

File: src/script/hooks/useAppSoftLock.ts [hunk 8]
```diff
@@ -20,34 +20,31 @@
...
 
-  const handleSoftLockActivation = useCallback(
-    async ({enrollmentConfig, identity}: {enrollmentConfig: EnrollmentConfig; identity?: WireIdentity}) => {
-      const isSoftLockEnabled = await shouldEnableSoftLock(enrollmentConfig, identity);
-
-      setSoftLockEnabled(isSoftLockEnabled);
-      callingRepository.setSoftLock(isSoftLockEnabled);
-      notificationRepository.setSoftLock(isSoftLockEnabled);
+  const handleDeviceStatusChange = useCallback(
+    ({status}: {status: 'valid' | 'locked'}) => {
```
Comment (przemvs) [inline src/script/hooks/useAppSoftLock.ts]: Hmm.. Maybe create type for status and reuse it? You have also this statuses there:
https://github.com/wireapp/wire-webapp/pull/16922/files#diff-080531c7467e54b71bd1e9e26a2d57a487984bf982fefb3f7afc152640f7da23R48

File: src/script/E2EIdentity/E2EIdentityEnrollment.ts (hunk 6 above)
Comment (atomrc) [reply src/script/E2EIdentity/E2EIdentityEnrollment.ts:62]: yeah Javascript privates 😎

File: src/script/E2EIdentity/E2EIdentityEnrollment.ts (hunk 7 above)
Comment (atomrc) [reply src/script/E2EIdentity/E2EIdentityEnrollment.ts]: true that 👌

File: src/script/hooks/useAppSoftLock.ts (hunk 8 above)
Comment (atomrc) [reply src/script/hooks/useAppSoftLock.ts]: fixed :) 
https://github.com/wireapp/wire-webapp/pull/16922/commits/037e05a352f17b7e1062c8e7aea1860e7c93bed9

File: src/script/E2EIdentity/SnoozableTimer/delay.ts [hunk 9]
```diff
@@ -17,60 +17,63 @@
...
+  if (identity?.status === MLSStatuses.EXPIRED) {
+    return {isSnoozable: false, firingDate: Date.now()};
   }
-  return identity.status === MLSStatuses.EXPIRED;
+
+  const expiryDate = identity?.certificate
+    ? Number(identity.notAfter) * TimeInMillis.SECOND
+    : deviceCreatedAt + teamGracePeriodDuration;
+
+  const isFirstEnrollment = !identity?.certificate;
```
Comment (przemvs) [inline src/script/E2EIdentity/SnoozableTimer/delay.ts]: WDYT ?

```suggestion
  const isFirstEnrollment = !identity?.certificate;
  const expiryDate = isFirstEnrollment
    ? deviceCreatedAt + teamGracePeriodDuration
    : Number(identity.notAfter) * TimeInMillis.SECOND;
```

**Review by przemvs — APPROVED** [review]

**Review by PatrykBuniX — APPROVED** [review]
Looks solid 🪨

**Pull request merged**